### PR TITLE
Introducing ItemGroup.allItems and similar default methods

### PR DESCRIPTION
**Pull request by jglick**
Have noticed (most recently in https://github.com/jenkinsci/active-choices-plugin/pull/17) that it is awkward to call a static utility method on what intuitively should be an instance method, which would be more readable and more discoverable with code completion.

### Proposed changelog entries

* API: Adding some default methods to `ItemGroup`.

@reviewbybees @jenkinsci/code-reviewers

### CODE REVIEW

File: core/src/main/java/hudson/model/ItemGroup.java [hunk 1]
```diff
@@ -88,4 +89,42 @@ default void onRenamed(T item, String oldName, String newName) throws IOExceptio
      * Internal method. Called by {@link Item}s when they are deleted by users.
      */
     void onDeleted(T item) throws IOException;
+
+    /**
+     * Gets all the {@link Item}s recursively in the {@link ItemGroup} tree
+     * and filter them by the given type.
+     * @since FIXME
+     */
+    default <T extends Item> List<T> getAllItems(Class<T> type) {
```
Comment (jglick) [inline core/src/main/java/hudson/model/ItemGroup.java]: This and others should really be `List<? extends T>` to signal immutability, but that would be source-incompatible for existing callers via `Jenkins`.

File: core/src/main/java/hudson/model/ItemGroup.java [hunk 2]
```diff
@@ -88,4 +89,42 @@ default void onRenamed(T item, String oldName, String newName) throws IOExceptio
...
+    /**
+     * Gets all the items unordered, lazily and recursively.
+     * @since FIXME
+     */
+    default Iterable<Item> allItems() {
+        return allItems(Item.class);
+    }
+
+    // TODO could delegate to allItems overload taking Authentication, but perhaps more useful to introduce a variant to perform preauth filtering using Predicate and check Item.READ afterwards
+    // or return a Stream<Item> and provide a Predicate<Item> public static Items.readable(), and see https://stackoverflow.com/q/22694884/12916 if you are looking for just one result
```
Comment (jglick) [inline core/src/main/java/hudson/model/ItemGroup.java]: Could easily write such an impl, but do not have a candidate caller handy to use as a downstream PR, and could find no cases of this idiom in core. (https://github.com/jenkinsci/active-choices-plugin/pull/17 could use it, I was just too lazy to file a PR based on a snapshot baseline for it. Maybe if someone expresses interest.)

File: core/src/test/java/hudson/model/ListViewTest.java [hunk 3]
```diff
@@ -24,6 +25,7 @@
     
     private interface ItemGroupOfNonTopLevelItem extends TopLevelItem, ItemGroup<Item> {}
 
+    @Ignore("TODO I am not smart enough to figure out what PowerMock is actually doing; whatever this was testing, better move to the test module and use JenkinsRule")
```
Comment (jglick) [inline core/src/test/java/hudson/model/ListViewTest.java]: Uh, yeah. Apply the transformation in this test suite and both cases fail with an opaque `NullPointerException`. Leave the suite alone and the other case passes but this one fails with a comparison error—unless you run just this case by itself, in which case it passes.

**Review by oleg-nenashev — APPROVED** [review]
🐝 and @reviewbybees done

Since we do lots of such renaming, we probably need to be aware of breaking method `@since` info in migrated methods (CC @daniel-beck). It is not a big deal though.

**Comment by daniel-beck** [comment]
Should this have deprecated now obsolete utility methods of `Items`?

**Comment by jglick** [comment]
> Should this have deprecated now obsolete utility methods of `Items`?

I considered it but decided to leave them as is, especially since the current `default` implementations continue to call them. No strong opinion.

**Pull request merged**